File: test/hardware/unit/test_mdu_unit.py
import cocotb
from cocotb.triggers import Timer, RisingEdge
from cocotb.clock import Clock
from test.driver import run_hardware_test
import random

# Operations
OP_MUL    = 0b000
OP_MULH   = 0b001
OP_MULHSU = 0b010
OP_MULHU  = 0b011
OP_DIV    = 0b100
OP_DIVU   = 0b101
OP_REM    = 0b110
OP_REMU   = 0b111

async def reset_dut(dut):
    dut.rst_n.value = 0
    dut.start.value = 0
    dut.operation.value = 0
    dut.operand_a.value = 0
    dut.operand_b.value = 0
    await Timer(20, units="ns")
    dut.rst_n.value = 1
    await RisingEdge(dut.clk)

async def run_operation(dut, op, a, b):
    dut.operation.value = op
    dut.operand_a.value = a
    dut.operand_b.value = b
    dut.start.value = 1
    await RisingEdge(dut.clk)
    dut.start.value = 0
    
    # Wait for ready
    while not dut.ready.value:
        await RisingEdge(dut.clk)
        
    return dut.result.value

@cocotb.test()
async def test_mdu_basic(dut):
    """
    Basic MDU Unit Test
    """
    clock = Clock(dut.clk, 10, units="ns")
    cocotb.start_soon(clock.start())
    
    await reset_dut(dut)
    
    # Test 1: MUL 10 * 5 = 50
    res = await run_operation(dut, OP_MUL, 10, 5)
    assert res.integer == 50, f"MUL 10*5 failed. Got {res.integer}"
    dut._log.info("MUL 10*5 passed")
    
    # Test 2: DIV 100 / 5 = 20
    res = await run_operation(dut, OP_DIV, 100, 5)
    assert res.integer == 20, f"DIV 100/5 failed. Got {res.integer}"
    dut._log.info("DIV 100/5 passed")
    
    # Test 3: REM 100 % 7 = 2
    res = await run_operation(dut, OP_REM, 100, 7)
    assert res.integer == 2, f"REM 100%7 failed. Got {res.integer}"
    dut._log.info("REM 100%7 passed")

    # Test 4: DIV by 0
    # RISC-V: x / 0 = -1 (all 1s)
    res = await run_operation(dut, OP_DIV, 100, 0)
    assert res.integer == -1 or res.integer == 4294967295, f"DIV by 0 failed. Got {res.integer}"
    dut._log.info("DIV by 0 passed")

    # Test 5: REM by 0
    # RISC-V: x % 0 = x
    res = await run_operation(dut, OP_REM, 123, 0)
    assert res.integer == 123, f"REM by 0 failed. Got {res.integer}"
    dut._log.info("REM by 0 passed")

    # Test 6: Signed MUL (Negative)
    # -10 * 5 = -50
    # -10 is 0xFFFFFFF6
    res = await run_operation(dut, OP_MUL, 0xFFFFFFF6, 5)
    # Result should be -50 (0xFFFFFFCE)
    # cocotb BinaryValue.integer is signed if we treat it so, but let's check unsigned representation or signed
    assert res.signed_integer == -50, f"MUL -10*5 failed. Got {res.signed_integer}"
    dut._log.info("MUL -10*5 passed")

    # Test 7: Signed DIV
    # -100 / 5 = -20
    res = await run_operation(dut, OP_DIV, 0xFFFFFF9C, 5)
    assert res.signed_integer == -20, f"DIV -100/5 failed. Got {res.signed_integer}"
    dut._log.info("DIV -100/5 passed")


def test_mdu_unit():
    run_hardware_test(
        module_name="test_mdu_unit",
        verilog_sources=["core/backend/mdu.v"],
        toplevel="mdu"
    )

if __name__ == "__main__":
    test_mdu_unit()

File: test/unit_test/test_mdu_unit.cpp
#define DOCTEST_CONFIG_IMPLEMENT_WITH_MAIN
#include "doctest.h"
#include "tb_base.h"
#include "Vmdu.h"

// MDU Operations
#define OP_MUL    0b000
#define OP_MULH   0b001
#define OP_MULHSU 0b010
#define OP_MULHU  0b011
#define OP_DIV    0b100
#define OP_DIVU   0b101
#define OP_REM    0b110
#define OP_REMU   0b111

/**
 * MDU (Multiply-Divide Unit) Testbench
 * Tests multiply and divide operations including signed/unsigned variants
 */
class MDUTestbench : public ClockedTestbench<Vmdu> {
public:
    MDUTestbench() : ClockedTestbench<Vmdu>(100, false) {
        dut->rst_n = 0;
        dut->start = 0;
        dut->operation = 0;
        dut->operand_a = 0;
        dut->operand_b = 0;
    }
    
    void set_clk(uint8_t value) override {
        dut->clk = value;
    }
    
    void reset() {
        dut->rst_n = 0;
        tick();
        tick();
        dut->rst_n = 1;
        tick();
    }
    
    uint32_t run_operation(uint8_t op, uint32_t a, uint32_t b) {
        dut->operation = op;
        dut->operand_a = a;
        dut->operand_b = b;
        dut->start = 1;
        tick();
        dut->start = 0;
        
        // Wait for ready
        int timeout = 100;
        while (!dut->ready && timeout-- > 0) {
            tick();
        }
        
        if (timeout <= 0) {
            throw std::runtime_error("MDU operation timeout");
        }
        
        return dut->result;
    }
    
    void test_multiply() {
        
        // MUL 10 * 5 = 50
        uint32_t res = run_operation(OP_MUL, 10, 5);
        CHECK(res == 50);
        
        // MUL -10 * 5 = -50
        res = run_operation(OP_MUL, 0xFFFFFFF6, 5);
        CHECK(static_cast<int32_t>(res) == -50);
        
        // MUL large numbers
        res = run_operation(OP_MUL, 1000, 2000);
        CHECK(res == 2000000);
    }
    
    void test_divide() {
        
        // DIV 100 / 5 = 20
        uint32_t res = run_operation(OP_DIV, 100, 5);
        CHECK(res == 20);
        
        // DIV -100 / 5 = -20
        res = run_operation(OP_DIV, 0xFFFFFF9C, 5);
        CHECK(static_cast<int32_t>(res) == -20);
        
        // DIV by 0 (should return -1 per RISC-V spec)
        res = run_operation(OP_DIV, 100, 0);
        CHECK(static_cast<int32_t>(res) == -1);
    }
    
    void test_remainder() {
        
        // REM 100 % 7 = 2
        uint32_t res = run_operation(OP_REM, 100, 7);
        CHECK(res == 2);
        
        // REM by 0 (should return dividend per RISC-V spec)
        res = run_operation(OP_REM, 123, 0);
        CHECK(res == 123);
    }
    
    void test_unsigned_operations() {
        
        // DIVU (unsigned divide)
        uint32_t res = run_operation(OP_DIVU, 0xFFFFFFFF, 2);
        CHECK(res == 0x7FFFFFFF);
        
        // REMU (unsigned remainder)
        res = run_operation(OP_REMU, 0xFFFFFFFF, 10);
        CHECK(res == 5);
    }
};

TEST_CASE("Mdu Unit") {
MDUTestbench tb;
        
        tb.reset();
        tb.test_multiply();
        tb.test_divide();
        tb.test_remainder();
        tb.test_unsigned_operations();
}
